llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp

```diff
@@ -2201,8 +2202,28 @@ UnPackOp::getTiledImplementation(OpBuilder &builder,
                                           dimAndTileMapping[dim]));
 
       inputIndices.push_back(firstCoord.quotient);
-      inputSizes.push_back(
-          add(sub(lastCoord.quotient, firstCoord.quotient), oneAttr));
+
+      // Get the upper bound because it could be an extract_slice case. The
+      // sizes are determined by loop bound and step, where loop bound is the
+      // size of output shape. In incomplete tile cases, the input could have
+      // larger shape, it is safe to extend the boundary because they are
+      // pre-padded.
+      FailureOr<int64_t> cstSize = linalg::getConstantUpperBoundForIndex(
+          getValueOrCreateConstantIndexOp(builder, loc, sizes[dim]));
+      Optional<int64_t> cstInnerSize =
+          getConstantIntValue(dimAndTileMapping[dim]);
+      if (!failed(cstSize) && cstInnerSize &&
+          cstSize.value() % cstInnerSize.value() == 0) {
+        // Assume that offset[dim] must be 0 or a multiple of sizes[dim]. This
+        // is true if the unpack op is untiled or generated by its tiling
+        // implementation. In this context, the input size is always
+        // (sizes[dim] / innerSize) in this case.
+        inputSizes.push_back(
+            builder.getIndexAttr(cstSize.value() / cstInnerSize.value()));
+      } else {
+        inputSizes.push_back(
+            add(sub(lastCoord.quotient, firstCoord.quotient), oneAttr));
+      }
```

Comment:
The affine inference is not smart enough for this case. I believe that it is correct, but I'm willing to have a deeper look with others.

the original form is something like `((d0 + 31) floor_div 32 - d0 + 1)`. I think it is just 1, but it can't be folded to 1.. After thinking a while, I decided to add my logic for it.

Reply:
Have been trying to understand this part, and read it multiple times and not able to actually figure out whats going on here.... Can you add some comments here with IR examples?